pkg/utils/utils.go

```diff
@@ -612,36 +614,70 @@ func GetRepositoryFromBackupRepository(backupRepository *backupdriverapi.BackupR
 }
 
 /*
- * Get the configuration to access the Supervisor namespace from the GuestCluster Pod
+ * Get the configuration to access the Supervisor namespace from the GuestCluster.
+ * This routine will be called only for guest cluster.
+ *
+ * The secret to access the Supervisor Cluster will be written in the backup driver
+ * namespace. The steps to get the Supervisor cluster config are:
+ * 1. Create the backup-driver namespace if it does not exist. This is a fixed namespace where
+ * where the para virt backup driver secret will be written.
+ * 2. Wait for the para virt backup driver secret to be written.
+ * 3. Get the supervisor cluster configuration from the cert and token in the secret.
+ * TODO: Handle update of the para virt backup driver secret
  */
-func SupervisorConfig(logger logrus.FieldLogger) (*rest.Config, string, error) {
-	token, err := ioutil.ReadFile(PvTokenLocation)
+func GetSupervisorConfig(guestConfig *rest.Config, logger logrus.FieldLogger) (*rest.Config, string, error) {
+	var err error
+	if guestConfig == nil {
+		guestConfig, err = rest.InClusterConfig()
+		if err != nil {
+			logger.WithError(err).Errorf("Failed to get k8s inClusterConfig")
+			return nil, "", errors.Wrap(err, "Could not retrieve in-cluster config")
+		}
+	}
+	clientset, err := kubernetes.NewForConfig(guestConfig)
 	if err != nil {
-		logger.WithError(err).Errorf("Failed to read the Para Virtual token from the credentials file %s", PvNamespaceLocation)
-		return nil, "", err
+		logger.WithError(err).Error("Failed to get k8s clientset from the given config")
+		return nil, "", errors.Wrap(err, "Could not retrieve in-cluster config client")
 	}
 
-	ns, err := ioutil.ReadFile(PvNamespaceLocation)
+	// Create Backup driver namespace if it does not exist
+	err = checkAndCreateNamespace(guestConfig, BackupDriverNamespace, logger)
 	if err != nil {
-		logger.WithError(err).Errorf("Failed to read the Para Virtual namespace from the credentials file %s", PvNamespaceLocation)
+		logger.WithError(err).Errorf("Failed to create namespace %s", BackupDriverNamespace)
 		return nil, "", err
 	}
 
-	tlsClientConfig := rest.TLSClientConfig{}
-
-	if _, err := certutil.NewPool(PvCrtLocation); err != nil {
-		klog.Errorf("Expected to load root CA config from %s, but got err: %v", PvCrtLocation, err)
+	// Check for the para virt secret. If it does not exist, wait for the secret to be written.
+	secretApis := clientset.CoreV1().Secrets(BackupDriverNamespace)
+	secret, err := secretApis.Get(context.TODO(), PvSecretName, metav1.GetOptions{})
+	if err == nil {
+		logger.Infof("Retrieved k8s secret %s in namespace %s", PvSecretName, BackupDriverNamespace)
 	} else {
-		tlsClientConfig.CAFile = PvCrtLocation
+		logger.Infof("Waiting for secret %s to be created in namespace %s", PvSecretName, BackupDriverNamespace)
+		ctx := context.Background()
+		secret, err = waitForPvSecret(ctx, clientset, BackupDriverNamespace, logger)
```

Comment:
What happens if the velero app operator fails to write the secret? do we wait here indefinitely?. I am wondering if this specific wait needs a timeout, maybe unnecessary for this change, but I guess having a todo would help.

Reply:
We cannot do much in the guest cluster backup driver without the secret. So we do wait indefinitely for the secret. We will hit this if we install the vsphere plugin in the guest without velero in the supervisor cluster.

Reply:
Fair enough, maybe add this piece of info as well in the logs, something like "ensure velero is installed in the supervisor", this could help the users self triage.